### ISSUE TITLE
Business rules test are broken

### ISSUE THREAD

**Issue by mohierf** [Alignak-monitoring/alignak#559]
The most simple tests with business rules does not pass. A simple Or between two services

-----

Configuration is:
```
define service{
  check_command                  check_service!ok
  host_name                      test_host_0
  service_description            db1
  use                            generic-service
}
define service{
  check_command                  check_service!ok
  host_name                      test_host_0
  service_description            db2
  use                            generic-service
}

define service{
  check_command                  bp_rule!test_host_0,db1|test_host_0,db2
  host_name                      test_host_0
  service_description            Simple_Or
  use                            generic-service
}
```

The BR associated to the service `Simple_Or` is built as a DependencyNode object that has a sons list (other DependencyNode objects). Each of its first level son has itself a Service object in its sons list.

So the BR object should contain a list of sons:
 - the first one which has a Service object  *db1* as unique son
 - the second one which has a Service object  *db2* as unique son

In the unit tests, when the configuration is parsed, this object tree is ok, but, a bit later in the unit tests, when one try to use this BR, the Service objects are no more present in the sons list ...

**Comment by mohierf** [comment]
With the unit test module I got the same Business Rule in the Arbiter and then in the Scheduler. The result is as following for the Arbiter:
```
BR arbiter: Op:'|' Val:'('2', '2', '2')' Sons:'[Op:'service' Val:'('0', '0', '0')' Sons:'[<Service host_name=u'test_host_0' desc=u'db1' name=u'generic-service' use=[u'generic-service'] />]' IsNot:'False',Op:'service' Val:'('0', '0', '0')' Sons:'[<Service host_name=u'test_host_0' desc=u'db2' name=u'generic-service' use=[u'generic-service'] />]' IsNot:'False']' IsNot:'False'
First son: Op:'service' Val:'(u'0', u'0', u'0')' Sons:'[Op:'None' Val:'('0', '0', '0')' Sons:'[]' IsNot:'False']' IsNot:'False'
Op:'None' Val:'('0', '0', '0')' Sons:'[]' IsNot:'False'
```
and in the scheduler
```
BR scheduler: Op:'|' Val:'(u'2', u'2', u'2')' Sons:'[Op:'service' Val:'(u'0', u'0', u'0')' Sons:'[Op:'None' Val:'('0', '0', '0')' Sons:'[]' IsNot:'False']' IsNot:'False',Op:'service' Val:'(u'0', u'0', u'0')' Sons:'[Op:'None' Val:'('0', '0', '0')' Sons:'[]' IsNot:'False']' IsNot:'False']' IsNot:'False'
```

The business rule got corrupted when the configuration was sent to the scheduler ...

**Comment by mohierf** [comment]
Got it!

The problem is due to a bad unserialization of the Service or Host objects when we recreate the objects after receiving a new configuration in the scheduler.

The `__init__` function of the DependencyNode object does not care if a `son` is not a DependencyNode ... it is the case when the `son` is an Host or a Service ...

**Comment by Seb-Solon** [comment]
Seems legit. The fix should be here : https://github.com/Alignak-monitoring/alignak/blob/develop/alignak/dependencynode.py#L87. I forgot to add the if when recreating. If you look at the serialize function the son list is created by serializing each member.

We have 3 options : 

* add a call to the serialize function (the one in misc) and unserialize when recreating object (edit unserialize and __init__). 
* Manually add info to know how to unserialize the son  (edit unserialize and __init__), similar to the first one
* Guess the type in __init__ based on keys in the serialized object (like host_name). Kinda hacky but reduce the patch  size

Pick what you think better, I have no other ideas on that

**Comment by mohierf** [comment]
Thanks, but the problem is mainly on the cyclic import ... because importing `alignak.objects.host` creates an import recursion in the DependencyNode  object initialization :(

**Comment by Seb-Solon** [comment]
Ok I is bigger than a simple serialization stuff. As we recreate object, they are not the same as the one recreated in self.hosts / services in the scheduler. So that are not linked. Therefore, the scheduler does not update them when they change state. 

We could have seen the test getting red when we refactored the alignak test class. Before all item were the same because we did not serialize / unserialize conf. 

We have to rework the whole dependencynode class to store uuid and not object. It should be similar to what we do with checks for example.

Working on it. I may have a fix tonight.
